From 8503604f2ca8dea05fd1e28a0ebee784aa47f2ba Mon Sep 17 00:00:00 2001
From: Koji Matsuoka <koji.matsuoka.xm@renesas.com>
Date: Tue, 7 Nov 2017 16:12:45 +0900
Subject: [PATCH 375/909] media: i2c: adv748x: Set default EDID value

commit 08749ab2ae94891c42eeb2eb711d86b97e107931 from
git://git.kernel.org/pub/scm/linux/kernel/git/horms/renesas-bsp.git

Depending on the input device, if video decoder device do not have
EDID, communication may fail and data transfer may not be possible.
Therefore, This patch set default EDID information as initial value
and solve the problem.

Signed-off-by: Koji Matsuoka <koji.matsuoka.xm@renesas.com>
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 drivers/media/i2c/adv748x/adv748x-hdmi.c | 74 ++++++++++++++++++++++++
 1 file changed, 74 insertions(+)

diff --git a/drivers/media/i2c/adv748x/adv748x-hdmi.c b/drivers/media/i2c/adv748x/adv748x-hdmi.c
index 17e38b8e0d14..6da8a20b92f4 100644
--- a/drivers/media/i2c/adv748x/adv748x-hdmi.c
+++ b/drivers/media/i2c/adv748x/adv748x-hdmi.c
@@ -518,6 +518,68 @@ static inline int adv748x_hdmi_edid_write_block(struct adv748x_hdmi *hdmi,
 	return err;
 }
 
+static const __u8 g_edid_data[256] = {
+	/* Header information(0-19th byte) */
+		0x00, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0x00,
+		0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+		0x01, 0x0C, 0x01, 0x03,
+	/* Basic display parameters(20-24th byte) */
+		0x80, 0x50, 0x2D, 0x78, 0x0A,
+	/* Chromaticity coordinates(25-34th byte) */
+		0x0D, 0xC9, 0xA0, 0x57, 0x47, 0x98, 0x27, 0x12,
+		0x48, 0x4C,
+	/* Established timing bitmap(35-37th byte) */
+		0x20, 0x00, 0x00,
+	/* Standard timing information(38-53th byte) */
+	/* Because they are unused, in this field, all values are 0101h. */
+		0x01, 0x01, 0x01, 0x01, 0x01, 0x01, 0x01, 0x01,
+		0x01, 0x01, 0x01, 0x01, 0x01, 0x01, 0x01, 0x01,
+	/* Descriptor blocks of Descriptor 1(54-71th byte) */
+		0x02, 0x3A, 0x80, 0x18, 0x71, 0x38, 0x2D, 0x40,
+		0x58, 0x2C, 0x45, 0x00, 0xDC, 0x0B, 0x11, 0x00,
+		0x00, 0x1E,
+	/* Descriptor blocks of Descriptor 2(72-89th byte) */
+		0x8C, 0x0A, 0xD0, 0x8A, 0x20, 0xE0, 0x2D, 0x10,
+		0x10, 0x3E, 0x96, 0x00, 0x58, 0xC2, 0x21, 0x00,
+		0x00, 0x18,
+	/* Descriptor blocks of Descriptor 3(90-107th byte) */
+		0x00, 0x00, 0x00, 0x7C, 0x00, 0x4D,
+		0x59, 0x20, 0x48, 0x44, 0x54, 0x56, 0x0A, 0x20,
+		0x20, 0x20, 0x20, 0x20,
+	/* Descriptor blocks of Descriptor 4(108-125th byte) */
+		0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x0A,
+		0x2E, 0x06, 0x00, 0x0A, 0x20, 0x20, 0x20, 0x20, 0x20, 0x20,
+	/* Number of extensions to follow(126th byte) */
+		0x01,
+	/* Checksum(127th byte) */
+		0x75,
+	/* CEA EDID Timing Extension Version 3 */
+		0x02, 0x03, 0x1E, 0x40, 0x83, 0x7F, 0x40, 0x05,
+		0x40, 0x48,
+		0x10,	 /* 1920x1080p @ 59.94/60 Hz */
+		0x05,	 /* 1920x1080i @ 59.94/60 Hz */
+		0x04,	 /* 1280x720p @ 59.94/60 Hz */
+		0x01,	 /* 640x480p @ 59.94/60 Hz */
+		0x02,	 /* 720x480p @ 59.94/60 Hz */
+		0x06,	 /* 720(1440)x480i @ 59.94/60 Hz */
+		0x15,	 /* 720(1440)x576i @ 50 Hz */
+		0x11,	 /* 720x576p @ 50 Hz */
+		0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+		0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+		0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+		0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+		0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+		0x00, 0x00, 0x20, 0x20, 0x20, 0x20, 0x20, 0x20,
+		0x20, 0x20, 0x20, 0x20, 0x20, 0x20, 0x00, 0x00,
+		0x00, 0xFF, 0x00, 0x0A, 0x20, 0x20, 0x20, 0x20,
+		0x20, 0x20, 0x20, 0x20, 0x20, 0x20, 0x20, 0x20,
+		0x00, 0x00, 0x00, 0x1B, 0x00, 0x0A, 0x20, 0x20,
+		0x20, 0x20, 0x20, 0x20, 0x20, 0x20, 0x20, 0x20,
+		0x20, 0x20, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+		0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+		0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0xD8,
+};
+
 static int adv748x_hdmi_set_edid(struct v4l2_subdev *sd, struct v4l2_edid *edid)
 {
 	struct adv748x_hdmi *hdmi = adv748x_sd_to_hdmi(sd);
@@ -732,6 +794,8 @@ int adv748x_hdmi_init(struct adv748x_hdmi *hdmi)
 	static const struct v4l2_dv_timings cea1280x720 =
 		V4L2_DV_BT_CEA_1280X720P30;
 	int ret;
+	int err;
+	struct v4l2_edid g_edid;
 
 	hdmi->timings = cea1280x720;
 
@@ -754,6 +818,16 @@ int adv748x_hdmi_init(struct adv748x_hdmi *hdmi)
 	if (ret)
 		goto err_free_media;
 
+	g_edid.pad = 0;
+	g_edid.start_block = 0;
+	g_edid.blocks = 2;
+	g_edid.edid = (__u8 *)g_edid_data;
+	err = adv748x_hdmi_set_edid(&hdmi->sd, &g_edid);
+	if (err < 0) {
+		v4l2_err(&hdmi->sd, "edid set error %d\n", err);
+		return err;
+	}
+
 	return 0;
 
 err_free_media:
-- 
2.17.1

